homeassistant/components/onvif/parsers.py

```diff
@@ -214,6 +214,35 @@ async def async_parse_cell_motion_detector(uid: str, msg) -> Event:
         return None
 
 
+@PARSERS.register("tns1:RuleEngine/LineDetector/Crossed")
+# pylint: disable=protected-access
+async def async_parse_line_crossed(uid: str, msg) -> Event:
+    """Handle parsing event message.
+
+    Topic: tns1:RuleEngine/LineDetector/Crossed
+    """
+    try:
+        video_source = ""
+        video_analytics = ""
+        rule = ""
+        for source in msg.Message._value_1.Source.SimpleItem:
+            if source.Name == "VideoSourceConfigurationToken":
+                video_source = source.Value
+            if source.Name == "VideoAnalyticsConfigurationToken":
+                video_analytics = source.Value
+            if source.Name == "Rule":
+                rule = source.Value
+
+        return Event(
+            f"{uid}_{msg.Topic._value_1}_{video_source}_{video_analytics}_{rule}",
```

Comment:
There will be a lot of underscores in a row if the default empty string values remain here.

Reply:
The [specification document](https://www.onvif.org/specs/srv/analytics/ONVIF-Analytics-Service-Spec.pdf) defines the schema for LineCrossed as follows.  Compliant devices should include all three of the above, and they are required to guarantee uniqueness.  All the additional guarding done here and in the other parsers are to maintain support for the non-compliant devices that a bajillion issues end up getting open for.
```xml
<tt:RuleDescription Name="tt:LineDetector">
    <tt:Parameters>
        <tt:SimpleItemDescription Name="Direction" Type="tt:Direction"/>
        <tt:ElementItemDescription Name="Segments" Type="tt:Polyline"/>
    </tt:Parameters>
    <tt:Messages>
        <tt:Source>
            <tt:SimpleItemDescription Name="VideoSourceConfigurationToken" Type="tt:ReferenceToken"/>
            <tt:SimpleItemDescription Name="VideoAnalyticsConfigurationToken" Type="tt:ReferenceToken"/>
            <tt:SimpleItemDescription Name="Rule" Type="xs:string"/>
        </tt:Source>
        <tt:Data>
            <tt:SimpleItemDescription Name="ObjectId" Type="xs:integer"/>
        </tt:Data>
        <tt:ParentTopic>tns1:RuleEngine/LineDetector/Crossed</tt:ParentTopic>
    </tt:Messages>
</tt:RuleDescription>
```

Reply:
Are the events still useful without the source information?

Reply:
It depends on the device really.  If it's coming from a single camera with only one line detector rule configured, then yes.  If it's from an NVR with multiple cameras, or there are multiple detection rules configured for the same device then that is where the source uniqueness allows someone to know exactly what rule was triggered. i.e. "Did they cross line A or line B?"

Reply:
Ok.